File: mlir/examples/dsp/SimpleBlocks/Output/TryDSPApps/BenchmarkTest/CCode/speakerIdentification.c
#include <stdio.h>
#include <math.h>

#define SAMPLE_RATE 1000
#define DURATION 1
#define SIGNAL_LENGTH (SAMPLE_RATE * DURATION)

// Function to generate voice signature (sinusoidal wave with two frequencies)
void generateVoiceSignature(double freq1, double freq2, double signal[], int length, int sample_rate) {
    for (int i = 0; i < length; i++) {
        double t = (double)i / sample_rate;
        signal[i] = sin(2 * M_PI * freq1 * t) + cos(2 * M_PI * freq2 * t);
    }
}

// Function to compute the dot product (correlation) between two signals
double correlate(double signal1[], double signal2[], int length) {
    double result = 0.0;
    for (int i = 0; i < length; i++) {
        result += signal1[i] * signal2[i];
    }
    return result;
}

// Function to find the index of the maximum value in an array
int argmax(double arr[], int length) {
    int max_index = 0;
    for (int i = 1; i < length; i++) {
        if (arr[i] > arr[max_index]) {
            max_index = i;
        }
    }
    return max_index;
}

// Main function
int main() {
    // Sample rate
    int sample_rate = SAMPLE_RATE;
    
    // Signal arrays
    double person1[SIGNAL_LENGTH], person2[SIGNAL_LENGTH], person3[SIGNAL_LENGTH], unknown_signal[SIGNAL_LENGTH];
    
    // Generate voice signatures for Alice, Bob, Charlie
    generateVoiceSignature(100, 200, person1, SIGNAL_LENGTH, sample_rate); // Alice
    generateVoiceSignature(150, 250, person2, SIGNAL_LENGTH, sample_rate); // Bob
    generateVoiceSignature(120, 180, person3, SIGNAL_LENGTH, sample_rate); // Charlie

    // Generate an unknown signal (Bob's signature in this case)
    generateVoiceSignature(150, 250, unknown_signal, SIGNAL_LENGTH, sample_rate);
    
    // Correlate unknown signal with each person's signature
    double max1 = correlate(person1, unknown_signal, SIGNAL_LENGTH);
    double max2 = correlate(person2, unknown_signal, SIGNAL_LENGTH);
    double max3 = correlate(person3, unknown_signal, SIGNAL_LENGTH);
    
    // Store correlation results
    double total_maxes[3] = {0, 0, 0};
    
    // Set correlation results
    total_maxes[0] = max1;
    total_maxes[1] = max2;
    total_maxes[2] = max3;
    
    // Find the index of the maximum correlation result
    int max_index = argmax(total_maxes, 3);
    
    // Output results
    printf("Max Index: %d\n", max_index);
    printf("Max Value: %f\n", total_maxes[max_index]);
    printf("Correlation with Alice: %f\n", max1);
    printf("Correlation with Bob: %f\n", max2);
    printf("Correlation with Charlie: %f\n", max3);
    
    return 0;
}

File: mlir/examples/dsp/SimpleBlocks/Output/TryDSPApps/BenchmarkTest/Matlab/speakerIdentification.m
SAMPLE_RATE = 1000;
INPUT_LENGTH = 61;
DURATION = INPUT_LENGTH / SAMPLE_RATE;
CORRELATION_LENGTH = 2 * INPUT_LENGTH - 1;

person1 = generateVoiceSignature(100, 200, INPUT_LENGTH, SAMPLE_RATE);
person2 = generateVoiceSignature(150, 250, INPUT_LENGTH, SAMPLE_RATE);
person3 = generateVoiceSignature(120, 180, INPUT_LENGTH, SAMPLE_RATE);
unknown_signal = generateVoiceSignature(150, 250, INPUT_LENGTH, SAMPLE_RATE);

correlation1 = correlate(person1, unknown_signal);
correlation2 = correlate(person2, unknown_signal);
correlation3 = correlate(person3, unknown_signal);

total_maxes = [max(correlation1), max(correlation2), max(correlation3)];

[max_value, max_index] = max(total_maxes);

temp2 = total_maxes(1);
temp3 = total_maxes(2);
temp4 = total_maxes(3);

fprintf('%d\t', max_index - 1);
fprintf('%.6f\t', temp2);
fprintf('%.6f\t', max_value);
fprintf('%.6f\t', temp3);
fprintf('%.6f %.6f %.6f\t', total_maxes);
fprintf('%.6f\n', temp4);

function signal = generateVoiceSignature(freq1, freq2, len, fs)
    t = (0:len-1) / fs;
    signal = sin(2 * pi * freq1 * t) + sin(2 * pi * freq2 * t);
end

function result = correlate(signal1, signal2)
    len = length(signal1);
    corr_len = 2 * len - 1;
    result = zeros(1, corr_len);
    for lag = 0:corr_len-1
        for i = 1:len
            j = lag - len + i;
            if j >= 1 && j <= len
                result(lag+1) = result(lag+1) + signal1(i) * signal2(j);
            end
        end
    end
end
